DistroLauncher/Ubuntu/InitTasks.cpp

```diff
@@ -0,0 +1,427 @@
+    wprintf(L"CheckInitTasks: failed with exit code: %u\n", exitCode_);
+  }
+}
+
+// Groups the pieces of information from a single user entry in the passwd database we care about.
+struct UserEntry {
+  std::string name;
+  ULONG uid = -1;
+};
+
+namespace fs = std::filesystem;
+fs::path wslConfPath() {
+  // init-once, lazily.
+  static fs::path etcWslConf =
+      fs::path{L"\\\\wsl.localhost"} / DistributionInfo::Name / "etc\\wsl.conf";
+  return etcWslConf;
+}
+
+void writeUserToWslConf(std::string_view name) {
+  auto oldPath = wslConfPath();
+  auto newPath = wslConfPath();
+  newPath.replace_extension(".new");
+
+  {
+    // Write the new file starting with the [user] section.
+    // ios::binary is needed otherwise '\r' will be added on Windows :(.
+    std::ofstream confNew{newPath, std::ios::binary};
+    confNew << "[user]\ndefault=" << name << '\n';
+
+    // Read existing content.
+    std::ifstream confOld{oldPath};
+    std::string original{std::istreambuf_iterator<char>(confOld), std::istreambuf_iterator<char>()};
+    if (!original.empty()) {
+      // and copy that to the new file.
+      std::copy(original.begin(), original.end(), std::ostreambuf_iterator<char>(confNew));
+    }
+  }
+
+  // Files are closed by now, we may replace the old file with the new one.
+  fs::rename(newPath, oldPath);
+}
+
+bool setDefaultUserViaWslApi(WslApiLoader& api, unsigned long uid) {
+  if (auto hr = api.WslConfigureDistribution(uid, WSL_DISTRIBUTION_FLAGS_DEFAULT); FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return false;
+  }
+  return true;
+}
+
+// Collects all users found in the NSS passwd database, sorted by UID.
+std::vector<UserEntry> getAllUsers(WslApiLoader& api);
+
+// Returns the defaultUser set in /etc/wsl.conf or the empty string if none is set.
+std::string defaultUserInWslConf();
+
+// Converts a multi-byte into a wide string.
+std::wstring str2wide(const std::string& str, UINT codePage = CP_THREAD_ACP);
+
+bool enforceDefaultUser(WslApiLoader& api) try {
+  auto users = getAllUsers(api);
+
+  if (users.empty()) {
+    // unexpectedly nothing to do
+    _putws(L"CheckInitTasks: couldn't find any users in NSS database\n");
+    return false;
+  }
+
+  if (auto name = defaultUserInWslConf(); !name.empty()) {
+    auto found = std::find_if(users.begin(), users.end(),
+                              [&name](const UserEntry& u) { return u.name == name; });
+    if (found != users.end()) {
+      return setDefaultUserViaWslApi(api, found->uid);
+    }
+    wprintf(
+        L"CheckInitTasks: user \"%s\" referred in /etc/wsl.conf "
+        L"was not found in the system, entry may be overwritten.\n",
+        str2wide(name).c_str());
+  }
+
+  if (auto uid = DistributionInfo::QueryUid(L""); uid != 0) {
+    auto found = std::find_if(users.begin(), users.end(),
+                              [&uid](const UserEntry& u) { return u.uid == uid; });
+    if (found != users.end()) {
+      writeUserToWslConf(found->name);
```

Comment:
We'd not need to do so because it would just be setting what is already set, i.e. if `DistributionInfo::QueryUid(L"")` returns something other than 0, it's because somehow the default user was set via WSL API or directly in the registry.